projectile.el

```diff
@@ -2598,6 +2598,46 @@ regular expression."
         (funcall ag-command search-term (projectile-project-root)))
     (error "Package 'ag' is not available")))
 
+;;;###autoload
+(defun projectile-find-all-occurrences (&optional arg)
+    "List all occurrences of the text being searched using completion.
+With a prefix argument ARG prompts you for a directory on which to run search in."
+    (interactive "P")
+    (let* ((text-to-search (read-string
+                            (projectile-prepend-project-name "Find all: ")
+                            (projectile-symbol-or-selection-at-point)))
+           (directory (if arg
+                          (file-name-as-directory
+                           (read-directory-name "Find in directory: "))
+                        (projectile-project-root)))
+           (file-to-search (projectile-files-with-string text-to-search directory)))
+      (projectile-completing-read
+       "All occurrences: "
+       (reduce #'append
+               (mapcar (lambda (file)
```

Comment:
Probably you can extract this and the action into some private functions which you can actually test and would be easier to understand.